openfisca_core/periods/date_unit.py

```diff
@@ -0,0 +1,259 @@
+            True
+
+            >>> "day" in DateUnit.ethereal
+            True
+
+            >>> "eternity" in DateUnit.ethereal
+            False
+
+        """
+
+        return DateUnit.DAY, DateUnit.MONTH, DateUnit.YEAR
+
+    @property
+    def eternal(self) -> Tuple[DateUnit, ...]:
+        """Creates a :obj:`tuple` of ``key`` with eternal items.
+
+        Returns:
+            tuple(str): A :obj:`tuple` containing the ``keys``.
+
+        Examples:
+            >>> DateUnit.eternal
+            (<DateUnit.ETERNITY(eternity)>,)
+
+            >>> DateUnit.ETERNITY in DateUnit.eternal
+            True
+
+            >>> "ETERNITY" in DateUnit.eternal
+            True
+
+            >>> "eternity" in DateUnit.eternal
+            True
+
+            >>> "day" in DateUnit.eternal
+            False
+
+        """
+
+        return (DateUnit.ETERNITY,)
+
+
+class DateUnit(Enum, metaclass = DateUnitMeta):
+    """The date units of a rule system.
+
+    Attributes:
+        index (:obj:`int`): The ``index`` of each item.
+        name (:obj:`str`): The ``name`` of each item.
+        value (tuple(str, int)): The ``value`` of each item.
+
+    Examples:
+        >>> repr(DateUnit)
+        "<enum 'DateUnit'>"
+
+        >>> repr(DateUnit.DAY)
+        '<DateUnit.DAY(day)>'
+
+        >>> str(DateUnit.DAY)
+        'DateUnit.DAY'
+
+        >>> dict([(DateUnit.DAY, DateUnit.DAY.value)])
+        {<DateUnit.DAY(day)>: 'day'}
+
+        >>> tuple(DateUnit)
+        (<DateUnit.WEEK_DAY(week_day)>, <DateUnit.WEEK(week)>, <DateUnit.DA...)
+
+        >>> len(DateUnit)
+        6
+
+        >>> DateUnit["DAY"]
+        <DateUnit.DAY(day)>
+
+        >>> DateUnit["day"]
+        <DateUnit.DAY(day)>
+
+        >>> DateUnit[DateUnit.DAY]
+        <DateUnit.DAY(day)>
+
+        >>> DateUnit("day")
+        <DateUnit.DAY(day)>
+
+        >>> DateUnit.DAY in DateUnit
+        True
+
+        >>> "DAY" in DateUnit
+        True
+
+        >>> "day" in DateUnit
+        True
+
+        >>> DateUnit.DAY == DateUnit.DAY
+        True
+
+        >>> "DAY" == DateUnit.DAY
+        True
+
+        >>> "day" == DateUnit.DAY
+        True
+
+        >>> DateUnit.DAY < DateUnit.DAY
+        False
+
+        >>> DateUnit.DAY > DateUnit.DAY
+        False
+
+        >>> DateUnit.DAY <= DateUnit.DAY
+        True
+
+        >>> DateUnit.DAY >= DateUnit.DAY
+        True
+
+        >>> "DAY" < DateUnit.DAY
+        False
+
+        >>> "DAY" > DateUnit.DAY
+        False
+
+        >>> "DAY" <= DateUnit.DAY
+        True
+
+        >>> "DAY" >= DateUnit.DAY
+        True
+
+        >>> "day" < DateUnit.DAY
+        False
+
+        >>> "day" > DateUnit.DAY
+        False
+
+        >>> "day" <= DateUnit.DAY
+        True
+
+        >>> "day" >= DateUnit.DAY
+        True
+
+        >>> DateUnit.DAY.index
+        2
+
+        >>> DateUnit.DAY.name
+        'DAY'
+
+        >>> DateUnit.DAY.value
+        'day'
+
+    .. versionadded:: 35.9.0
+
+    """
```

Comment:
This has two benefits:

1. We consolidate the definition of date units in just one place, instead of using literals everywhere.

2. We can, at some point, use `DateUnit.Day.index` instead of `day` or `DateUnit.value`. So to be able to pass from:

```py
if period.unit == "eternity"
period.unit in ["eternirty", ...]
```

To :

```py
allowed_periods = DateUnit.DAY | DateUnit.MONTH

definition_period = [0, 1, 0, 0]

isperiod = definition_period[allowed_periods]
```